src/core/ext/transport/chaotic_good/client/chaotic_good_connector.cc

```diff
@@ -0,0 +1,339 @@
+      });
+}
+
+auto ChaoticGoodConnector::DataEndpointWriteSettingsFrame(
+    std::shared_ptr<ChaoticGoodConnector> self) {
+  GPR_ASSERT(self->data_endpoint_ != nullptr);
+  return TrySeq([self = self]() mutable {
+    // Serialize setting frame.
+    SettingsFrame frame;
+    // frame.header set connectiion_type: control
+    ClientMetadataHandle metadata =
+        GetContext<Arena>()->MakePooled<ClientMetadata>(GetContext<Arena>());
+    metadata->Set(ChaoticGoodConnectionTypeMetadata(),
+                  Slice::FromCopiedString("data"));
+    metadata->Set(ChaoticGoodConnectionIdMetadata(),
+                  self->connection_id_.Ref());
+    metadata->Set(ChaoticGoodDataAlignmentMetadata(), self->data_alignment_);
+    frame.headers = std::move(metadata);
+    auto write_buffer = frame.Serialize(&self->hpack_compressor_);
+    return self->data_endpoint_->Write(std::move(write_buffer.control));
+  });
+}
+
+auto ChaoticGoodConnector::WaitForDataEndpointSetup(
+    std::shared_ptr<ChaoticGoodConnector> self) {
+  // Data endpoint on_connect callback.
+  grpc_event_engine::experimental::EventEngine::OnConnectCallback
+      on_data_endpoint_connect =
+          [self = self](absl::StatusOr<std::unique_ptr<EventEngine::Endpoint>>
+                            endpoint) mutable {
+            self->data_endpoint_latch_->Set(std::make_shared<PromiseEndpoint>(
+                std::move(endpoint.value()), SliceBuffer()));
+            auto cb = self->wait_for_data_endpoint_callback_->MakeCallback();
+            // Wake up wait_for_data_endpoint_callback_.
+            cb();
+          };
+  self->event_engine_->Connect(
+      std::move(on_data_endpoint_connect), *self->resolved_addr_,
+      self->ee_config_,
+      ResourceQuota::Default()->memory_quota()->CreateMemoryAllocator(
+          "data_endpoint_connection"),
+      EventEngine::Duration(self->timeout_));
+
+  return TrySeq(
+      self->wait_for_data_endpoint_callback_->MakeWaitPromise(),
+      Race(TrySeq(self->data_endpoint_latch_->Wait(),
+                  [self = self](
+                      std::shared_ptr<PromiseEndpoint> data_endpoint) mutable {
+                    self->data_endpoint_.swap(data_endpoint);
+                    return TrySeq(
+                        DataEndpointWriteSettingsFrame(self),
+                        DataEndpointReadSettingsFrame(self),
+                        []() -> absl::Status { return absl::OkStatus(); });
+                  }),
+           TrySeq(Sleep(Timestamp::Now() + Duration::Seconds(self->timeout_)),
+                  []() -> absl::Status {
+                    return absl::DeadlineExceededError(
+                        "Data endpoint connect deadline excced.");
+                  })));
+}
+
+auto ChaoticGoodConnector::ControlEndpointReadSettingsFrame(
+    std::shared_ptr<ChaoticGoodConnector> self) {
+  GPR_ASSERT(self->control_endpoint_ != nullptr);
+  return TrySeq(
+      []() {
```

Comment:
I think this was an attempt to make sure that the `ReadSlice()` was called from the activities context.

Probably a better way of writing this would be:

```
return [self]() {
  return TrySeq(
    self->control_endpoint_->ReadSlice(...),
    [self](Slice slice) mutable {
      ...
```

i.e. return a factory for this TrySeq which can be instantiated by the TrySeq caller of this function